File: genetic_algorithms/word_mutation.rb
module Selection
  class << self
    def random_char(chars)
      chars.sample 
    end

    def mutate(parent,rate,chars)
      parent.each_char.collect { |char| rand <= rate ? random_char(chars) : char }.join
    end

    def mutation_rate(candidate,target)
      1.0 - Math.exp( -(100.0 - Selection.w(candidate,target)).fdiv(400))
    end

    def w(candidate,target) # Calculates fitness using the P0 = e^-x Poisson relationship
      Math.exp(candidate.chars.zip(target.chars).inject(0) { |a,b| a + (b[0].ord - b[1].ord).abs }.fdiv(-10))*100
    end

    def avg_mutation_rate(rates)
      rates.inject(:+).fdiv(rates.length)
    end
  end
end

module GeneticOutput
  class << self
    def edit_sentence(target,starting)
      starting.length > target.length ? (return starting.split("").sample(target.length).join("")) : (return starting + ("-"*(target.length-starting.length)))
    end

    def log(iteration,rate,target,fitness,parent,last=nil)
      last.nil? ? (puts "%4d %.2f %5.1f %s" % [iteration,rate,fitness,parent]) : (puts "\nGeneration: %4d\nAverage Mutation Rate: %.2f\nFitness: %5.1f\n%s" % [iteration,rate,fitness,parent])
    end
  end
end

class WordMutation
  attr_accessor :first_gen,:last_gen,:gen_count,:rates
  attr_reader   :target,:num_copies,:search_chars
  def initialize(target,copies=100)
    @target       = target
    @num_copies   = copies
    @rates        = []
    @last_gen     = ""
    @gen_count    = 1
    @search_chars = [("A".."Z").to_a,("a".."z").to_a," "].flatten
  end

  def set_parent(starting=nil)
    starting.nil? ? @first_gen = search_chars.sample(target.length).join("") : @first_gen = GeneticOutput.edit_sentence(target,starting)
  end

  def run_mutations
    until first_gen == target
      rates << Selection.mutation_rate(first_gen,target)
      reset_generation(gen_count,rates.last) until last_gen == first_gen
      @first_gen = (1..num_copies).map { Selection.mutate(first_gen,rates.last,search_chars) }.push(first_gen).max_by { |n| Selection.w(n,target) }
      @gen_count += 1
    end
  end

  def reset_generation(iteration,rate)
    GeneticOutput.log(iteration,rate,target,Selection.w(first_gen,target),first_gen)
    @last_gen = first_gen
  end

  def final_output
    GeneticOutput.log(gen_count,Selection.avg_mutation_rate(rates),target,Selection.w(first_gen,target),first_gen,"last")
  end
end

sentence = WordMutation.new("This is a test phrase")
sentence.set_parent
sentence.run_mutations
sentence.final_output

File: genetic_algorithms/word_mutation.exs
defmodule Rand do
  def init do
    << a :: 32, b :: 32, c :: 32 >> = :crypto.strong_rand_bytes(12)
    :random.seed(a,b,c)
  end 

  def num do
    init
    :random.uniform
  end

  def char(list) do
    Enum.at(list, :random.uniform(length(list)) - 1)
  end
end

defmodule Log do
  def show(offspring,i) do
    IO.puts "Generation: #{i}, Offspring: #{offspring}"
  end

  def found([target|i]) do
    IO.puts "#{target} found in #{i} iterations"
  end
end

defmodule Evolution do
  def select(target) do
    chars = (?A..?Z) |> Enum.to_list() |> List.insert_at(0, 32)
    (1..String.length(target))
      |> Enum.map(fn _-> Rand.char(chars) end)
      |> mutate(String.to_char_list(target),0,[],chars)
      |> Log.found()
  end

  def mutate(parent,target,i,_,_)     when target == parent, do: [parent|i]
  def mutate(parent,target,i,_,chars) when target != parent do
    w = fitness(parent,target)
    prev = reproduce(target,parent,w,0,mu_rate(w),chars)
    if w < fitness(prev,target) do
      parent = prev
      Log.show(parent,i)
    end
    mutate(parent,target,i+1,prev,chars)
  end

  def reproduce(target,parent,_,_,rate,chars) do
    (1..100) 
      |> Enum.to_list() 
      |> Stream.map(fn _-> mutation(parent,rate,chars) end)
      |> List.insert_at(0, parent) 
      |> Enum.max_by(fn n -> fitness(n,target) end)
  end

  def fitness(t,r) do
    (0..length(t)-1) 
      |> Stream.map(fn n -> abs(Enum.at(t,n) - Enum.at(r,n)) end) 
      |> Enum.reduce(fn a,b -> a + b end)
      |> calc()
  end

  def mutation(p,r,chars) do
    (0..length(p)-1) 
      |> Stream.map(fn n -> Enum.at(p,n) end)
      |> Enum.map(fn n -> if Rand.num <= r, do: Rand.char(chars), else: n end)
  end

  def calc(sum),  do: 100 * :math.exp(sum/-10)
  def mu_rate(n), do: 1   - :math.exp(-(100-n)/400)
end

Evolution.select("THIS IS A TEST PHRASE")